pkg/ring/lifecycler.go

```diff
@@ -631,6 +897,18 @@ func (i *Lifecycler) changeState(ctx context.Context, state IngesterState) error
 
 	level.Info(util.Logger).Log("msg", "changing ingester state from", "old_state", currState, "new_state", state)
 	i.setState(state)
+
+	// If we're not actively leaving the ring and incremental join is disabled,
+	// we'll set all of our tokens to the same state we just changed into.
+	//
+	// Likewise, we'll do the same if we are actively leaving the ring and
+	// incremental leave is disabled.
+	if !i.cfg.JoinIncrementalTransfer && state != LEAVING {
+		i.setTokensState(state)
+	} else if !i.cfg.LeaveIncrementalTransfer && state == LEAVING {
+		i.setTokensState(state)
+	}
```

Comment:
A better way of handling this is to conditionally enable whether or not the ingester is using stateful tokens.

Reply:
I'm backpedaling on this. I tried implementing it but it's more complicated to follow than having this here.